src/util/queryNodeMatchers.ts

```diff
@@ -0,0 +1,147 @@
+import { Position, Range } from "vscode";
+import { SyntaxNode, Query, Language, QueryCapture } from "web-tree-sitter";
+import {
+  NodeMatcher,
+  NodeMatcherValue,
+  SelectionExtractor,
+  SelectionWithEditor,
+} from "../typings/Types";
+import { simpleSelectionExtractor } from "./nodeSelectors";
+
+export function defaultMatcher(
+  scopeType: string,
+  searchScopePresent: boolean,
+  scopeQuery: string,
+  selector: SelectionExtractor = simpleSelectionExtractor
+): NodeMatcher {
+  let query: Query;
+
+  return (
+    selection: SelectionWithEditor,
+    node: SyntaxNode,
+    siblings: boolean | undefined
+  ): NodeMatcherValue[] | null => {
+    if (!query) {
+      query = getQuery(node, scopeQuery);
+    }
+
+    const captures = extractCaptures(
+      selection,
+      node.tree.rootNode,
+      query,
+      scopeType
+    );
+
+    if (captures.length === 0) {
+      return null;
+    }
+    let capture: QueryCapture;
+    capture = selectCaptureByRange(captures, selection);
+    if (siblings) {
+      let captures: QueryCapture[];
+      if (searchScopePresent) {
+        throw new Error("searchScope based queries are not implemented.");
+      } else {
+        captures = findBySiblingsByParent(
+          capture.node,
+          selection,
+          query,
+          scopeType
+        );
+      }
+
+      return captures!.map((c) => {
+        return {
+          node: c.node,
+          selection: selector(selection.editor, c.node),
+        };
+      });
+    } else {
+      return [
+        {
+          node: capture!.node,
+          selection: selector(selection.editor, capture!.node),
+        },
+      ];
+    }
+  };
+}
+
+function extractCaptures(
+  selection: SelectionWithEditor,
+  root: SyntaxNode,
+  query: Query,
+  scopeType: string
+) {
+  const startPoint = generatePointFromSelection(selection, "start");
+  const endPoint = generatePointFromSelection(selection, "end");
+
+  return query.captures(root, startPoint, endPoint).filter((capture) => {
+    return capture.name === scopeType;
+  });
+}
+
+function selectCaptureByRange(
+  captures: QueryCapture[],
+  selection: SelectionWithEditor
+) {
+  let capture: QueryCapture;
+  captures.forEach((c) => {
+    const range: Range = new Range(
+      new Position(c.node.startPosition.row, c.node.startPosition.column),
+      new Position(c.node.endPosition.row, c.node.endPosition.column)
+    );
+
+    const hasIntersection = selection.selection.intersection(range);
```

Comment:
What if selection is empty? I think adjacent empty selection should count as intersecting

Reply:
Yep, good point. I think that this should be handled via `extractCaptures` however. By this point, we'll not have any captures, if we didn't expand the selection. I do think it makes sense to:

1. Consider making the widest selection within `extractCaptures`. Perhaps expand `startPoint` and `endPoint` to the preceding and following non-whitespace characters. `query.captures(root, startPoint, endPoint)`
2. Rely on the initial selection, as we do above, initially. 
3. Expand the selection outwards, one way at a time, prioritizing the `active` position over the `anchor` position.

I'm debating if I want to act on the third point by calling `selectCaptureByRange` more than once or expanding within this loop.

Reply:
ah ok so the tree-sitter function `query.captures(root, startPoint, endPoint)` won't return the node if it is adjacent to an empty `startPoint, endPoint` range?  That's a bummer.  I would lean towards your option 1) above (expanding left and right just past first non-whitespace char)